compiler/src/dotty/tools/dotc/ast/tpd.scala

```diff
@@ -40,8 +40,11 @@ object tpd extends Trees.Instance[Type] with TypedTreeInfo {
   def Super(qual: Tree, mixName: TypeName, inConstrCall: Boolean, mixinClass: Symbol = NoSymbol)(implicit ctx: Context): Super =
     Super(qual, if (mixName.isEmpty) untpd.EmptyTypeIdent else untpd.Ident(mixName), inConstrCall, mixinClass)
 
-  def Apply(fn: Tree, args: List[Tree])(implicit ctx: Context): Apply =
+  def Apply(fn: Tree, args: List[Tree])(implicit ctx: Context): Apply = {
+    assert(!fn.isInstanceOf[Block])
```

Comment:
Just a suggestion, but you could be even more radical and assert what fn can be instead of what it cannot be, is `assert(fn.isInstanceOf[RefTree] || fn.isInstanceOf[GenericApply])` enough ?

Reply:
Same for `TypeApply`.